IDEasy broken and always failing out of project (IDE_HOME)

### Expected behavior

As a IDEasy, I want to use the tool to get help or create a project so that it can be of any use.

### Actual behavior

```
$ ide help
An unexpected error occurred!
We are sorry for the inconvenience.
Please check the error below, resolve it and try again.
If the error is not on your end (network connectivity, lack of permissions, etc.) please file a bug:
https://github.com/devonfw/ide/issues/new?assignees=&labels=bug&projects=&template=bug.md&title=IllegalStateException%3A+Running+command+%27C%3A%5CProgram+Files%5CGit%5Cmingw64%5Cbin%5Cgit.exe%27+with+arguments+%27rev-parse%27+%27HEAD%27+failed%3A+Running+command+%27C%3A%5CProgram+Files%5CGit%5Cmingw64%5Cbin%5Cgit.exe%27+with+arguments+%27rev-parse%27+%27HEAD%27+failed+with+exit+code+128%21
java.lang.IllegalStateException: Running command 'C:\Program Files\Git\mingw64\bin\git.exe' with arguments 'rev-parse' 'HEAD' failed: Running command 'C:\Program Files\Git\mingw64\bin\git.exe' with arguments 'rev-parse' 'HEAD' failed with exit code 128!
        at com.devonfw.tools.ide.process.ProcessContextImpl.run(ProcessContextImpl.java:173)
        at com.devonfw.tools.ide.context.GitContextImpl.isRepositoryUpdateAvailable(GitContextImpl.java:133)
        at com.devonfw.tools.ide.context.AbstractIdeContext.applyAndRun(AbstractIdeContext.java:861)
        at com.devonfw.tools.ide.context.AbstractIdeContext.run(AbstractIdeContext.java:779)
        at com.devonfw.tools.ide.cli.Ideasy.runOrThrow(Ideasy.java:89)
        at com.devonfw.tools.ide.cli.Ideasy.run(Ideasy.java:52)
        at com.devonfw.tools.ide.cli.Ideasy.main(Ideasy.java:29)
        at java.base@21.0.2/java.lang.invoke.LambdaForm$DMH/sa346b79c.invokeStaticInit(LambdaForm$DMH)
Caused by: com.devonfw.tools.ide.cli.CliException: Running command 'C:\Program Files\Git\mingw64\bin\git.exe' with arguments 'rev-parse' 'HEAD' failed with exit code 128!
        at com.devonfw.tools.ide.process.ProcessContextImpl.performLogOnError(ProcessContextImpl.java:289)
        at com.devonfw.tools.ide.process.ProcessContextImpl.run(ProcessContextImpl.java:164)
        ... 7 more
```

### Steps to reproduce (bug) / Use Case of feature request (enhancement)

1. Download latest SNAPSHOT release of IDEasy - e.g. [ide-cli-2024.08.002-beta-20240830.023446-8-windows-x64.tar.gz](https://s01.oss.sonatype.org/content/repositories/snapshots/com/devonfw/tools/IDEasy/ide-cli/2024.08.002-beta-SNAPSHOT/ide-cli-2024.08.002-beta-20240830.023446-8-windows-x64.tar.gz)
2. Run `setup.bat`
3. Run `ide help`

### Related/Dependent Issues

Story: #161
PR: #467

### Comments/Hints:

Bug is here:
https://github.com/devonfw/IDEasy/blob/1e62bd2a4a949f9875cac16118bf5108940eb431/cli/src/main/java/com/devonfw/tools/ide/context/AbstractIdeContext.java#L861C55-L861C70

We first need to check if `settingsPath` is `null` (because already `IDE_HOME` is `null` because we are not inside an IDEasy project) otherwise none of this code must be executed.
Since you first need to create a project with `ide create` command that is not working because of this bug, we have created a critical blocker bug.

### Affected version:

* OS: any
* IDEasy: 2024.08.002-beta-20240830.023446-8

--- a/cli/src/main/java/com/devonfw/tools/ide/context/AbstractIdeContext.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/context/AbstractIdeContext.java
@@ -858,9 +858,11 @@ public abstract class AbstractIdeContext implements IdeContext {
           debug(getMessageIdeHomeFound());
         }
       }
-      if (getGitContext().isRepositoryUpdateAvailable(getSettingsPath()) ||
-          (getGitContext().fetchIfNeeded(getSettingsPath()) && getGitContext().isRepositoryUpdateAvailable(getSettingsPath()))) {
-        info("Updates are available for the settings repository. If you want to pull the latest changes, call ide update.");
+      if (this.settingsPath != null) {
+        if (getGitContext().isRepositoryUpdateAvailable(this.settingsPath) ||
+            (getGitContext().fetchIfNeeded(this.settingsPath) && getGitContext().isRepositoryUpdateAvailable(this.settingsPath))) {
+          info("Updates are available for the settings repository. If you want to pull the latest changes, call ide update.");
+        }
       }
       cmd.run();
     } else {

--- a/cli/src/main/java/com/devonfw/tools/ide/context/GitContext.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/context/GitContext.java
@@ -44,13 +44,13 @@ public interface GitContext {
    * This method checks the last modified time of the `FETCH_HEAD` file in the `.git` directory to determine if a fetch is needed based on a predefined
    * threshold. If updates are available in the remote repository, it logs an information message prompting the user to pull the latest changes.
    *
+   * @param targetRepository the {@link Path} to the target folder where the git repository is located. It contains the `.git` subfolder.
    * @param remoteName the name of the remote repository, e.g., "origin".
    * @param branch the name of the branch to check for updates.
-   * @param targetRepository the {@link Path} to the target folder where the git repository is located. It contains the `.git` subfolder.
    * @return {@code true} if updates were detected after fetching from the remote repository, indicating that the local repository is behind the remote.
    *     {@code false} if no updates were detected or if no fetching was performed (e.g., the cache threshold was not met or the context is offline)
    */
-  boolean fetchIfNeeded(String remoteName, String branch, Path targetRepository);
+  boolean fetchIfNeeded(Path targetRepository, String remoteName, String branch);
 
   /**
    * Checks if there are updates available for the Git repository in the specified target folder by comparing the local commit hash with the remote commit

--- a/cli/src/main/java/com/devonfw/tools/ide/context/GitContextImpl.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/context/GitContextImpl.java
@@ -88,11 +88,11 @@ public class GitContextImpl implements GitContext {
     String remoteName = remoteAndBranch[0];
     String branch = remoteAndBranch[1];
 
-    return fetchIfNeeded(remoteName, branch, targetRepository);
+    return fetchIfNeeded(targetRepository, remoteName, branch);
   }
 
   @Override
-  public boolean fetchIfNeeded(String remoteName, String branch, Path targetRepository) {
+  public boolean fetchIfNeeded(Path targetRepository, String remoteName, String branch) {
 
     if (!context.isOnline()) {
       return false;

--- a/cli/src/main/java/com/devonfw/tools/ide/context/GitContextImpl.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/context/GitContextImpl.java
@@ -94,7 +94,7 @@ public class GitContextImpl implements GitContext {
   @Override
   public boolean fetchIfNeeded(Path targetRepository, String remoteName, String branch) {
 
-    if (!context.isOnline()) {
+    if (this.context.isOffline()) {
       return false;
     }
 

--- a/cli/src/test/java/com/devonfw/tools/ide/context/GitContextMock.java
+++ b/cli/src/test/java/com/devonfw/tools/ide/context/GitContextMock.java
@@ -61,7 +61,7 @@ public class GitContextMock implements GitContext {
   }
 
   @Override
-  public boolean fetchIfNeeded(String remoteName, String branch, Path targetRepository) {
+  public boolean fetchIfNeeded(Path targetRepository, String remoteName, String branch) {
 
     return false;
   }

--- a/cli/src/test/java/com/devonfw/tools/ide/context/GitContextTest.java
+++ b/cli/src/test/java/com/devonfw/tools/ide/context/GitContextTest.java
@@ -177,7 +177,7 @@ public class GitContextTest extends AbstractIdeContextTest {
     GitContext gitContext = new GitContextImpl(context);
 
     // act
-    gitContext.fetchIfNeeded(repoUrl, remoteName, tempDir);
+    gitContext.fetchIfNeeded(tempDir, repoUrl, remoteName);
 
     // assert
     // Since the context is offline, no actions should be performed
@@ -207,7 +207,7 @@ public class GitContextTest extends AbstractIdeContextTest {
     Files.setLastModifiedTime(fetchHead, FileTime.fromMillis(System.currentTimeMillis()));
 
     // act
-    gitContext.fetchIfNeeded(repoUrl, remoteName, tempDir);
+    gitContext.fetchIfNeeded(tempDir, repoUrl, remoteName);
 
     // assert
     // Since FETCH_HEAD is recent, no fetch should occur

